### DISCUSSION TITLE
Questions on Multithreading Performance in HOOMD-blue Simulations of Active Brownian Particles

### DISCUSSION THREAD

**Discussion by idop11** [glotzerlab/hoomd-blue#1720] [Support] [answered]
Hello,

I have two questions regarding the use of multiple threads in HOOMD-blue version 4.4.1:

1. Are any additional changes required to use multiple threads? I understand that many functions in the package may not support multiprocessing independently, while others do.

2. I have attached a code ("ABPs.txt") that simulates the dynamics of active Brownian particles. The simulation initially minimizes the system's energy using the FIRE algorithm before running the actual dynamics. In this simulation, the interaction between particles is modeled with the DPDConservative force, and each particle is an active particle with a propulsion speed of $v_0$ and undergoes orientation changes due to rotational thermal noise. I observed that setting ```cpu = hoomd.device.CPU(num_cpu_threads=4)``` and ```cpu = hoomd.device.CPU()``` results in the same execution time for each run. Could this be due to certain functions that I assumed could run independently (e.g., calculating the DPDConservative force for each particle at each iteration) but actually cannot, or is there another reason?

By the way, if there is anything in my code that is incorrect or does not accurately represent the desired dynamics, I would appreciate feedback. I am new to the HOOMD-blue package.

Thanks!

[ABPs.txt](https://github.com/glotzerlab/hoomd-blue/files/14232829/ABPs.txt)

**Comment by joaander** [comment, marked as answer]
As [stated in the documentation](https://hoomd-blue.readthedocs.io/en/v4.4.1/features.html#threading), 
> threading support in HOOMD-blue is very limited and only applies to implicit depletants in [hpmc.integrate.HPMCIntegrator](https://hoomd-blue.readthedocs.io/en/v4.4.1/module-hpmc-integrate.html#hoomd.hpmc.integrate.HPMCIntegrator), and [hpmc.pair.user.CPPPotentialUnion](https://hoomd-blue.readthedocs.io/en/v4.4.1/module-hpmc-pair-user.html#hoomd.hpmc.pair.user.CPPPotentialUnion)

I recommend that you use 1 GPU or MPI parallelization on an appropriate number of CPU cores.
